src/bitwidth.rs

```diff
@@ -11,68 +14,129 @@ use crate::{
+///     // because this function returns an error if `x.width()` is larger than the target
+///     // width.
+///     x.sign_extend(width)?;
+///
+///     // Since error handling is involved no matter what, we have made the function
+///     // signature accept `target_width: W` where
+///     // `W: TryInto<BitWidth, Error = E>, crate::Error: From<E>`. This means that any
+///     // type with an impl for `TryInto<BitWidth>` can be used as an argument. There is
+///     // an `impl TryFrom<usize> for BitWidth`, so a plain `usize` can be entered into
+///     // the function, and the function will handle both `BitWidth` invariant checking
+///     // and its own invariants.
+///     x.sign_extend(100)?;
+///
+///     Ok(x)
+/// }
+///
+/// example_function(64).unwrap();
+/// ```
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct BitWidth(usize);
+pub struct BitWidth(NonZeroUsize);
 
-//  ===========================================================================
-///  Constructors
-/// ===========================================================================
-impl BitWidth {
-    /// Creates a `BitWidth` that represents a bit-width of `1` bit.
-    #[inline]
-    pub fn w1() -> Self {
-        BitWidth(1)
-    }
-
-    /// Creates a `BitWidth` that represents a bit-width of `8` bits.
-    #[inline]
-    pub fn w8() -> Self {
-        BitWidth(8)
-    }
-
-    /// Creates a `BitWidth` that represents a bit-width of `16` bits.
-    #[inline]
-    pub fn w16() -> Self {
-        BitWidth(16)
-    }
-
-    /// Creates a `BitWidth` that represents a bit-width of `32` bits.
-    #[inline]
-    pub fn w32() -> Self {
-        BitWidth(32)
+impl From<NonZeroUsize> for BitWidth {
+    /// Creates a `BitWidth` from the given `NonZeroUsize`.
+    fn from(width: NonZeroUsize) -> Self {
+        BitWidth(width)
     }
+}
 
-    /// Creates a `BitWidth` that represents a bit-width of `64` bits.
-    #[inline]
-    pub fn w64() -> Self {
-        BitWidth(64)
-    }
-
-    /// Creates a `BitWidth` that represents a bit-width of `128` bits.
-    #[inline]
-    pub fn w128() -> Self {
-        BitWidth(128)
-    }
+impl TryFrom<usize> for BitWidth {
+    type Error = crate::Error;
 
     /// Creates a `BitWidth` from the given `usize`.
     ///
     /// # Errors
     ///
     /// - If the given `width` is equal to zero.
-    pub fn new(width: usize) -> Result<Self> {
-        if width == 0 {
-            return Err(Error::invalid_zero_bitwidth())
+    fn try_from(width: usize) -> Result<BitWidth> {
+        match NonZeroUsize::new(width) {
+            Some(bitwidth) => Ok(BitWidth(bitwidth)),
+            None => Err(crate::Error::invalid_zero_bitwidth()),
+        }
+    }
+}
+
+/// Convenience free function for converting a `usize` to a `BitWidth`.
+///
+/// # Panics
+///
+/// If `width == 0`, this function will panic
+pub const fn bw(width: usize) -> BitWidth {
```

Comment:
I like the approach of having a utility free function that does exactly this conversion job. However, we currently shouldn't provide it as `const` function if it requires us to make use of nightly-only Rust features. Instead we could define it as non-`const` function and make it `const` once all required features are stabilized.

Reply:
There are techniques to work around the current limitations of Rust's const features. However, those techniques each come with their drawbacks. If you are still interested in doing some research you could start with the `typenum` or `generic-array` crates.

Reply:
I will make it non-const and open an issue for making it const once the needed features are stabilized